File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/LocksReleasedAllPaths.ReaderWriterLockSlim.cs
﻿using System;
using System.Threading;

namespace ReaderWriterLockSlim_Type
{
    class Program
    {
        private bool condition;
        private ReaderWriterLockSlim readerWriterLockSlim = new ReaderWriterLockSlim();

        public void Method1()
        {
            readerWriterLockSlim.EnterReadLock(); // Noncompliant
            if (condition)
            {
                readerWriterLockSlim.ExitReadLock();
            }
        }

        public void Method2()
        {
            readerWriterLockSlim.EnterWriteLock(); // Noncompliant
            if (condition)
            {
                readerWriterLockSlim.ExitWriteLock();
            }
        }

        public void Method3()
        {
            readerWriterLockSlim.EnterUpgradeableReadLock(); // Noncompliant
            if (condition)
            {
                readerWriterLockSlim.ExitUpgradeableReadLock();
            }
        }

        public void Method4()
        {
            if (readerWriterLockSlim.TryEnterReadLock(42)) // Noncompliant
            {
            }
            else
            {
                readerWriterLockSlim.ExitReadLock();
            }
        }

        public void Method5()
        {
            if (readerWriterLockSlim.TryEnterReadLock(new TimeSpan(42))) // Noncompliant
            {
            }
            else
            {
                readerWriterLockSlim.ExitReadLock();
            }
        }

        public void Method6()
        {
            if (readerWriterLockSlim.TryEnterWriteLock(42)) // Noncompliant
            {
            }
            else
            {
                readerWriterLockSlim.ExitWriteLock();
            }
        }

        public void Method7()
        {
            if (readerWriterLockSlim.TryEnterWriteLock(new TimeSpan(42))) // Noncompliant
            {
            }
            else
            {
                readerWriterLockSlim.ExitWriteLock();
            }
        }

        public void Method8()
        {
            if (readerWriterLockSlim.TryEnterUpgradeableReadLock(42)) // Noncompliant
            {
            }
            else
            {
                readerWriterLockSlim.ExitReadLock();
            }
        }

        public void Method9()
        {
            if (readerWriterLockSlim.TryEnterUpgradeableReadLock(new TimeSpan(42))) // Noncompliant
            {
            }
            else
            {
                readerWriterLockSlim.ExitReadLock();
            }
        }


        public void Method10()
        {
            try
            {
                readerWriterLockSlim.EnterUpgradeableReadLock();
                readerWriterLockSlim.EnterWriteLock(); // Compliant
                if (condition)
                {
                    readerWriterLockSlim.ExitWriteLock();
                }
            }
            catch (Exception)
            {
                throw;
            }
            finally
            {
                readerWriterLockSlim.ExitUpgradeableReadLock();
            }
        }

        public void Method11(string arg)
        {
            try
            {
                readerWriterLockSlim.EnterUpgradeableReadLock();
                readerWriterLockSlim.EnterWriteLock();
                try
                {
                    Console.WriteLine(arg.Length);
                }
                catch (Exception)
                {

                    throw;
                }
                finally
                {
                    readerWriterLockSlim.ExitWriteLock();
                }
            }
            catch (Exception)
            {

                throw;
            }
            finally
            {
                readerWriterLockSlim.ExitUpgradeableReadLock();
            }
        }

        public void Method12()
        {
            readerWriterLockSlim.EnterReadLock(); // Compliant
            readerWriterLockSlim.ExitReadLock();
        }

        public void Method13()
        {
            readerWriterLockSlim.EnterReadLock(); // Compliant, this rule doesn't care if it was released with correct API
            readerWriterLockSlim.ExitWriteLock();
        }

        public void WrongOrder()
        {
            readerWriterLockSlim.ExitReadLock();
            readerWriterLockSlim.EnterReadLock(); // Noncompliant

            var a = new ReaderWriterLockSlim();
            a.ExitWriteLock();
            a.EnterWriteLock(); // Noncompliant

            var b = new ReaderWriterLockSlim();
            b.ExitUpgradeableReadLock();
            b.TryEnterReadLock(1); // Noncompliant

            var c = new ReaderWriterLockSlim();
            c.ExitReadLock();
            c.TryEnterWriteLock(1); // Noncompliant

            var d = new ReaderWriterLockSlim();
            d.ExitReadLock();
            d.EnterUpgradeableReadLock(); // Noncompliant

            var e = new ReaderWriterLockSlim();
            e.ExitReadLock();
            e.TryEnterUpgradeableReadLock(1); // Noncompliant
        }

        public void Method14()
        {
            readerWriterLockSlim.EnterReadLock(); // Noncompliant, this rule doesn't care if it was released with correct API
            if (condition)
            {
                readerWriterLockSlim.ExitWriteLock();
            }
        }

        public void Method15()
        {
            if (readerWriterLockSlim.TryEnterReadLock(42)) // Noncompliant FP
            {
                readerWriterLockSlim.ExitReadLock();
            }
        }

        public void EarlyExit()
        {
            if (readerWriterLockSlim.TryEnterReadLock(0) == false) // Noncompliant FP - https://github.com/SonarSource/sonar-dotnet/issues/5415
            {
                return;
            }

            try
            {
            }
            finally
            {
                readerWriterLockSlim.ExitReadLock();
            }
        }

        public void EarlyExit_UnaryOperator()
        {
            if (!readerWriterLockSlim.TryEnterReadLock(0)) // Noncompliant FP - https://github.com/SonarSource/sonar-dotnet/issues/5415
            {
                return;
            }

            try
            {
            }
            finally
            {
                readerWriterLockSlim.ExitReadLock();
            }
        }

        public void Throw_Finally(object param)
        {
            readerWriterLockSlim.EnterReadLock(); // Noncompliant FP, there are multiple ocurrences on peach. Should be handled by throw implementation.
            try
            {
                if (param == null)
                    throw new ObjectDisposedException("");

                return;
            }
            finally
            {
                readerWriterLockSlim.ExitReadLock();
            }
        }

        public void IsReadLockHeld()
        {
            readerWriterLockSlim.EnterReadLock(); // Noncompliant FP, https://github.com/SonarSource/sonar-dotnet/issues/5416
            if (readerWriterLockSlim.IsReadLockHeld)
            {
                readerWriterLockSlim.ExitReadLock();
            }
        }

        public void IsWriteLockHeld()
        {
            readerWriterLockSlim.EnterWriteLock(); // Noncompliant FP, https://github.com/SonarSource/sonar-dotnet/issues/5416
            if (readerWriterLockSlim.IsWriteLockHeld)
            {
                readerWriterLockSlim.ExitWriteLock();
            }
        }
    }
}

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/LocksReleasedAllPaths.ReaderWriterLockSlim.vb
﻿Imports System.Threading

Namespace ReaderWriterLockSlim_Type

    Class Program

        Private Condition As Boolean
        Private rwLockSlim As New ReaderWriterLockSlim()

        Public Sub Method1()
            rwLockSlim.EnterReadLock() ' Noncompliant
            If Condition Then rwLockSlim.ExitReadLock()
        End Sub

        Public Sub Method2()
            rwLockSlim.EnterWriteLock() ' Noncompliant
            If Condition Then rwLockSlim.ExitWriteLock()
        End Sub

        Public Sub Method3()
            rwLockSlim.EnterUpgradeableReadLock() ' Noncompliant
            If Condition Then rwLockSlim.ExitUpgradeableReadLock()
        End Sub

        Public Sub Method4()
            If rwLockSlim.TryEnterReadLock(42) Then ' Noncompliant
            Else
                rwLockSlim.ExitReadLock()
            End If
        End Sub

        Public Sub Method5()
            If rwLockSlim.TryEnterReadLock(New TimeSpan(42)) Then ' Noncompliant
            Else
                rwLockSlim.ExitReadLock()
            End If
        End Sub

        Public Sub Method6()
            If rwLockSlim.TryEnterWriteLock(42) Then ' Noncompliant
            Else
                rwLockSlim.ExitWriteLock()
            End If
        End Sub

        Public Sub Method7()
            If rwLockSlim.TryEnterWriteLock(New TimeSpan(42)) Then ' Noncompliant
            Else
                rwLockSlim.ExitWriteLock()
            End If
        End Sub

        Public Sub Method8()
            If rwLockSlim.TryEnterUpgradeableReadLock(42) Then ' Noncompliant
            Else
                rwLockSlim.ExitReadLock()
            End If
        End Sub

        Public Sub Method9()
            If rwLockSlim.TryEnterUpgradeableReadLock(New TimeSpan(42)) Then ' Noncompliant
            Else
                rwLockSlim.ExitReadLock()
            End If
        End Sub


        Public Sub Method10()
            Try
                rwLockSlim.EnterUpgradeableReadLock()
                rwLockSlim.EnterWriteLock() ' Compliant
                If Condition Then rwLockSlim.ExitWriteLock()
            Catch ex As Exception
                Throw
            Finally
                rwLockSlim.ExitUpgradeableReadLock()
            End Try
        End Sub

        Public Sub Method11(Arg As String)
            Try
                rwLockSlim.EnterUpgradeableReadLock()
                rwLockSlim.EnterWriteLock()
                Try
                    Console.WriteLine(Arg.Length)
                Catch ex As Exception
                    Throw
                Finally
                    rwLockSlim.ExitWriteLock()
                End Try
            Catch ex As Exception
                Throw
            Finally
                rwLockSlim.ExitUpgradeableReadLock()
            End Try
        End Sub

        Public Sub Method12()
            rwLockSlim.EnterReadLock() ' Compliant
            rwLockSlim.ExitReadLock()
        End Sub

        Public Sub Method13()
            rwLockSlim.EnterReadLock() ' Compliant, this rule doesn't care if it was released with correct API
            rwLockSlim.ExitWriteLock()
        End Sub

        Public Sub WrongOrder()
            rwLockSlim.ExitReadLock()
            rwLockSlim.EnterReadLock()  ' Noncompliant

            Dim a = New ReaderWriterLockSlim()
            a.ExitWriteLock()
            a.EnterWriteLock()          ' Noncompliant

            Dim b = New ReaderWriterLockSlim()
            b.ExitUpgradeableReadLock()
            b.TryEnterReadLock(1)       ' Noncompliant

            Dim c = New ReaderWriterLockSlim()
            c.ExitReadLock()
            c.TryEnterWriteLock(1)      ' Noncompliant

            Dim d = New ReaderWriterLockSlim()
            d.ExitReadLock()
            d.EnterUpgradeableReadLock()        ' Noncompliant

            Dim e = New ReaderWriterLockSlim()
            e.ExitReadLock()
            e.TryEnterUpgradeableReadLock(1)    ' Noncompliant
        End Sub

        Public Sub Method14()
            rwLockSlim.EnterReadLock() ' Noncompliant, this rule doesn't care if it was released with correct API
            If Condition Then rwLockSlim.ExitWriteLock()
        End Sub

        Public Sub Method15()
            If rwLockSlim.TryEnterReadLock(42) Then ' Noncompliant FP
                rwLockSlim.ExitReadLock()
            End If
        End Sub

    End Class

End Namespace
